Add WebGL2 support

davepagurek opened the pull request:
Resolves https://github.com/processing/p5.js/issues/2536

### Changes
- Defaults p5 to WebGL2 mode
- Adds global `p5.disableWebGL2` to force WebGL1 mode
- Adds `webglVersion` to graphics and the main canvas to read which version was picked
- Updates p5 code using WebGL1-specific things to handle both environments:
  - The font shader relied on a WebGL1 extension to use `fwidth`. This exists by default in WebGL2, but only if you upgrade the glsl version from 2 to 3. I added some glsl macros which will be the correct glsl 2 or 3 keyword depending on the environment.
  - Two blend modes, LIGHTEST and DARKEST, use a WebGL1 extension that is now default in WebGL2. It now tries to use the default versions (e.g. `MIN`) before falling back to the extension version (e.g. `MIN_EXT`)

### Screenshots of the change:
No visual change should occur!

I have a test sketch here https://editor.p5js.org/davepagurek/sketches/-XNYgrz_z mirroring the example added to the docs. Commenting/uncommenting the `p5.disableWebGL2` line will make the sketch show "WebGL 1" or "WebGL 2" on canvas, and also demonstrates the font shader working in both environments.


#### PR Checklist


- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

aferriss commented on src/core/main.js:
```diff
@@ -729,6 +729,36 @@ p5.instance = null;
+ * and will fall back to a WebGL1 context if WebGL2 is not available. If you
+ * want to always use a WebGL1 context, you can set `p5.disableWebGL2 = true`.
+ *
+ * @property {Boolean} disableWebGL2
```
could we do this in the setAttributes function instead? It feels against the grain of existing p5 patterns to be setting something on the p5 object before the sketch is created. Something like:
```
setAttributes('version' 1);
// or
setAttributes({version: 2});
```
Is there a reason we wouldn't do it this way?

aferriss commented on src/core/environment.js:
```diff
@@ -369,6 +369,35 @@ p5.prototype.noCursor = function() {
+ * function preload() {
+ *   myFont = loadFont('assets/inconsolata.otf');
+ * }
+ * function setup() {
```
Is there any other way to illustrate the difference between webGL1 vs 2 aside from swapping out a string based on the version?

davepagurek replied:
Maybe we could swap out two different shaders based on the webgl version, where one uses GLSL 2 and the other 3? It might be a bit verbose as an example though.

aferriss replied:
I was hoping there would be a simple visual example. I like the idea of showing the different shaders, maybe it would help people to see the differences side by side like that.

davepagurek replied:
Most of the features seem to be behind-the-scenes changes that you'd see in the code more than in the output. Looking at this https://webgl2fundamentals.org/webgl/lessons/webgl2-whats-new.html maybe the biggest visual change would be support for repeated textures of arbitrary sizes? Where in v1 it would fall back to clamp but in v2 would repeat?

aferriss reviewed (changes requested):
This is great, thanks so much for pulling it together @davepagurek !

I left a few notes but one other thing I think we should include are some tests. Can you add a few for each new piece of the api, and maybe some others  to make sure that font rendering is still working in both versions. Also maybe one more to make sure webGL versions are working with p5.graphics?

aferriss commented:
Also, I think there are a couple more spots in the library that use webGL1 extensions that might need some special handling.

There's the [float textures in p5.Texture](https://github.com/processing/p5.js/blob/af0fc0d85d7b2fa00463941fe214ad9e78c0b696/src/webgl/p5.Texture.js#L51)

Also this [uint extension in the drawElements function](https://github.com/processing/p5.js/blob/6420c4e3a13af115275c884466317ed274cc15bf/src/webgl/p5.RendererGL.Retained.js#L203)

aferriss commented:
I think in webGL two we can also do away with the power of two texture size checks as well in p5.Texture, especially if people are trying to use repeat wrap functions
https://github.com/processing/p5.js/blob/af0fc0d85d7b2fa00463941fe214ad9e78c0b696/src/webgl/p5.Texture.js#L350

aferriss commented:
Definitely not a blocker for this PR but at some point in the future it might be good for us to try and switch over the rendering to use VAO's if possible, as described in [this article](https://webgl2fundamentals.org/webgl/lessons/webgl1-to-webgl2.html)

davepagurek commented on src/core/environment.js:
```diff
@@ -369,6 +369,35 @@ p5.prototype.noCursor = function() {
+ * WebGL2 is not available.
+ *
+ * `webglVersion` will always be either `WEBGL2` or `WEBGL`.
+ * @property {String} webglVersion
```
I think many of the function ones are functions because they use the pattern where `frameRate(value)` is a setter while `frameRate()` is a getter. It makes sense from an OOP standpoint for `getTargetFrameRate()` to be a function so that you can't accidentally set the value, but it's also a bit of a deviation from seemingly the rest of p5, where read-only values are just properties (`mouseX`, `pmouseX`, `frameCount`, etc.)

I think if we make the webgl version settable via `setAttributes`, having its value exposed as a property would probably be most consistent then? But I don't think any other attributes are later readable (via a documented API anyway), so whatever we choose to do here can maybe set the precedent.

> What will this value return if the p5 instance is not in webGL mode?

Right now it's `WEBGL2`, which is probably confusing. Do you think it should be something like `undefined` or `P2D`?

aferriss replied:
I think returning the current renderer at `P2D` makes sense to me. 

Lets go ahead with returning the exposed value as a property then. I suppose most of the other ones aren't available simply because it probably isn't very useful information to have. I guess if we wanted to we could implement a `getAttributes()` function that would return either a specific key or the attributes object of the current renderer. 

Probably better to keep things simple for now and we can revisit those other attributes at a later time if the issue arises.

davepagurek commented:
Thanks for catching those other WebGL1 extension cases @aferriss! I've updated those parts to do nothing if using WebGL2, added some tests, and switched to using `setAttributes({ version: 1 })` as the way to request a WebGL1 canvas. For now it's still using `webglVersion` as a property, but I can update this some more if we decide on a better way to do that. Also still using a very simple example in the docs, but let me know if you think a shader example wouldn't be too much (maybe using `dFdx` and `dFdy` in GLSL3 to do a custom normal shader?)

aferriss commented on src/core/environment.js:
```diff
@@ -369,6 +369,36 @@ p5.prototype.noCursor = function() {
 };
 
+/**
+ * If the was created in WebGL mode, then `weglVersion` will indicate which
```
nit— should probably say: "If the sketch was created...."

aferriss commented on test/unit/webgl/p5.RendererGL.js:
```diff
@@ -23,6 +23,45 @@ suite('p5.RendererGL', function() {
+        HTMLCanvasElement.prototype.getContext = prevGetContext;
+      });
+
+      test('should return WEBGL1', function() {
```
I think you added this test twice?

davepagurek replied:
oops good catch, I left the `setAttributes({ version: 1})` in there by accident! I'll take that line out again so that it tests to make sure you still end up with webgl1 even when you don't request it if the browser doesn't support webgl2 (mocked in the `suite` this is nested in.)

aferriss reviewed (changes requested):
Just a couple small changes and then I think we'll be good to roll!

Can we add documentation for the new setAttributes property here: https://github.com/processing/p5.js/blob/main/src/webgl/p5.RendererGL.js#L405

Is there anywhere else in the docs that we can indicate that webGL mode uses webGL2 by default now? Maybe a note on https://p5js.org/reference/#/p5/createCanvas and https://p5js.org/reference/#/p5/createGraphics ?

Can you link the inconsolata font from the [manual test examples](https://github.com/processing/p5.js/blob/main/test/manual-test-examples/p5.Font/Inconsolata-Bold.ttf) instead of adding another copy?

davepagurek commented:
Updated those tests and docs! I have an idea for a better visual example of the difference between webgl1 and 2 so I'll make one more update after work with that.

davepagurek commented:
Updated the webglVersion example to show a visual difference in the capabilities of WebGL2!

![webglversion-example](https://user-images.githubusercontent.com/5315059/224190739-e7af2ebd-1df0-489e-985e-0872b69f8600.gif)

aferriss reviewed (approved):
lgtm! 🚀

Qianqianye commented:
Oh wow, this is exciting! Thanks @davepagurek and @aferriss for working on it. Hi @stalgiag, does this PR looks good to you to be merged?

stalgiag commented on src/webgl/p5.RendererGL.js:
```diff
@@ -1251,16 +1282,39 @@ p5.RendererGL.prototype._getLineShader = function() {
-      defaultShaders.fontFrag
+      this._webGL2CompatibilityPrefix('vert', 'mediump') +
+        defaultShaders.fontVert,
+      this._webGL2CompatibilityPrefix('frag', 'mediump') +
```
I am curious if this float precision is always guaranteed to work well on a device that supports WebGL2. Not sure but it is irrelevant because this design will make adjustments to precision based on the device easier in the future.

stalgiag commented on src/webgl/p5.Texture.js:
```diff
@@ -351,7 +354,10 @@ p5.Texture.prototype.setWrapMode = function(wrapX, wrapY) {
-    if (widthPowerOfTwo && heightPowerOfTwo) {
+    if (
+      this._renderer.webglVersion === constants.WEBGL2 ||
+      (widthPowerOfTwo && heightPowerOfTwo)
```
Might be worth rolling this into a simple function that returns a boolean just to improve readability. I find the nested conditionals and multiline conditions to be a challenge to read in this section (and many other sections of the GL code). Not a new problem with this PR by any means but one worth keeping in mind.

stalgiag reviewed (approved):
Amazing! This is so exciting! Thanks for the hard work @aferriss and @davepagurek!
